lib/bulk/common.js

```diff
@@ -1,5 +1,6 @@
 'use strict';
 
+const provided = require('../provided_promise');
```

Comment:
I have a mild preference for mongoose's naming of `PromiseProvider`, `provided` feels a little under specific and maybe doesn't quite describe what we're using here. You need to get the "provided Promise" from the "Promise provider":

```
const Promise = PromiseProvider.get();
return new Promise((resolve, reject) => {});
```

Reply:
I considered this, I originally copied their code however that would mean with every use of `Promise` we'd need to define it, or we'd use `ProverPromise.get().resolve()` or `new ProverPromise.get()((reject, resolve) => {})` which is very messy. We wouldn't be able to use `const Promise = require('../provided_promise').get();` because that provides a stale version of the stored promise. The alternative I created allows use of an exported getter. I am also against use of overwriting global promise `const Promise`, It's as if we did `const Array`.

Reply:
> I am also against use of overwriting global promise `const Promise`

Can you expand a bit on this? I thought that was the standard way of using a custom promise library, e.g. `const Promise = require('bluebird');`

Reply:
@emadum I would advise against it because `const Promise = require('bluebird');` is a shim.  Meant to add non-existing functionality to the runtime or replace it. However we use native promises in the codebase along with the user provided promise library. We need to be able to access either or at any time within the same scope. If ALL promises were to use the user-provided library I would find it slightly more acceptable. However even when I write code for personal projects with `bluebird` I personally do this `const Bluebird = require('bluebird');` because all the additional methods aren't apart of native `Promise` like `Promise.try`, `Promise.reduce`, `Promise.props` is can lead to confusion.

Reply:
I'm fine with defining `Promise` everywhere we use it. Given that the majority of our use of `Promise` is behind the `maybePromise` and `executeOperation` helpers, I wouldn't expect it to be too cumbersome to add it to the remaining places (which should probably be using the helpers in the first place!). I also agree with your assessment that we need to define it at the function level, because the provided implementation might change at runtime. 

One thing we haven't considered here is that we currently (maybe "magically", read: in an undocumented way) support changing the `promiseLibrary` at the operation level. You might need to do something like:
```
const Promise = PromiseProvider.get(options); // resolve the provided promise with one optionally provided at the function level `options`
return new Promise((resolve, reject) => {});
```

Reply:
I'm ok declaring it (not using a getter) but I still have fears when declaring the library as `Promise`.

I'm still a bit worried about the case where a function uses both native promises as well as user-defined. [Here's a case of this.](https://github.com/mongodb/node-mongodb-native/blob/master/lib/change_stream.js#L539-L557) It may be a mistake, perhaps both should be using the `promiseLibrary`. Let's assume it is not, defining `const Promise` to the user-promise at the top of the file is potentially not intended, and defining it below this native call is not great as well. I would still advocate for not declaring `Promise` in favor of:

```js
const PromiseLibrary = providedPromise.get()
```

This allows for more readability and not needing to know the entire scope to ensure what's going on. 

Another example is in the future if we see a PR diff and all we see is:

```
...
return Promsie.resolve()
...
```

We wouldn't need to check to see if the commiter overwritten promise or not. We'd know that is a native promise. Conversely we'd know that they declared `PromiseLibrary` and are using it here:

```
...
return PromiseLibrary.resolve()
...
```

I guess what I am looking for is clarity on native `Promise` and the user-defined `PromsieLibrary` or `PromiseProvided`. I want to easily be able to scan a file and know which is which, as well as be able to do searches for what's using a native `Promise` and what's using the user-defined.

Reply:
We could enforce this on the linter level, e.g. the [no-native rule in eslint-plugin-promise](https://github.com/xjamundx/eslint-plugin-promise/blob/HEAD/docs/rules/no-native.md).

Reply:
@emadum the linter solution SGTM